kernel/src/cpu/x86/apic.rs

```diff
@@ -1,16 +1,99 @@
 // SPDX-License-Identifier: MIT
 //
 // Copyright (c) Microsoft Corporation
+// Copyright (c) SUSE LLC
 //
 // Author: Jon Lange <jlange@microsoft.com>
+// Author: Joerg Roedel <jroedel@suse.de>
 
-pub const APIC_MSR_EOI: u32 = 0x80B;
-pub const APIC_MSR_ISR: u32 = 0x810;
-pub const APIC_MSR_ICR: u32 = 0x830;
+use crate::cpu::msr::{read_msr, write_msr};
 
-// Returns the MSR offset and bitmask to identify a specific vector in an
-// APIC register (IRR, ISR, or TMR).
+/// End-of-Interrupt register MSR offset
+pub const MSR_X2APIC_EOI: u32 = 0x80B;
+/// Spurious-Interrupt-Register MSR offset
+pub const MSR_X2APIC_SPIV: u32 = 0x80F;
+/// Interrupt-Service-Register base MSR offset
+pub const MSR_X2APIC_ISR: u32 = 0x810;
+/// Interrupt-Control-Register register MSR offset
+pub const MSR_X2APIC_ICR: u32 = 0x830;
+
+const MSR_APIC_BASE: u32 = 0x1B;
+const APIC_ENABLE_MASK: u64 = 0x800;
+const APIC_X2_ENABLE_MASK: u64 = 0x400;
+
+// SPIV bits
+const APIC_SPIV_VECTOR_MASK: u64 = (1u64 << 8) - 1;
+const APIC_SPIV_SW_ENABLE_MASK: u64 = 1 << 8;
+
+/// Get the MSR offset relative to a bitmap base MSR and the mask for the MSR
+/// value to check for a specific vector bit being set in IRR, ISR, or TMR.
+///
+/// # Returns
+///
+/// A `(u32, u32)` tuple with the MSR offset as the first and the vector
+/// bitmask as the second value.
 pub fn apic_register_bit(vector: usize) -> (u32, u32) {
     let index: u8 = vector as u8;
     ((index >> 5) as u32, 1 << (index & 0x1F))
 }
+
+/// Enables the X2APIC by setting the AE and EXTD bits in the APIC base address
+/// register.
+pub fn x2apic_enable() {
+    // Enable X2APIC mode.
+    let apic_base = read_msr(MSR_APIC_BASE);
+    let apic_base_x2_enabled = apic_base | APIC_ENABLE_MASK | APIC_X2_ENABLE_MASK;
+    if apic_base != apic_base_x2_enabled {
+        write_msr(MSR_APIC_BASE, apic_base_x2_enabled);
+    }
+    // Set SW-enable in SPIV to enable IRQ delivery
+    x2apic_sw_enable();
+}
+
+/// Send an End-of-Interrupt notification to the X2APIC.
+pub fn x2apic_eoi() {
+    write_msr(MSR_X2APIC_EOI, 0);
+}
```

Comment:
If we keep EOI in the platform abstraction, then this could be written like this and avoid #VC in the SNP case.

```suggestion
pub fn x2apic_eoi(wrmsr: FnOnce<(u32, u64)>) {
    wrmsr(MSR_X2APIC_EOI, 0);
}
```

Passing a closure to perform the WRMSR permits each platform to implement this optimally without requiring any virtualization exception.

Reply:
As per our discussion, lets solve this in PR #639